src/discover-addons.js

```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+function discoverAddonsRecursive(current, name, foundAddons = []) {
+  let currentAddons = current.addons || [];
+  let found = currentAddons.find(addon => addon.name === name);
+  if (found) {
+    foundAddons.push(found);
+  }
+
+  currentAddons.forEach(addon =>
+    discoverAddonsRecursive(addon, name, foundAddons)
+  );
+
+  return foundAddons;
+}
+
+module.exports = function(project, name) {
+  return discoverAddonsRecursive(project, name);
```

Comment:
the built-in functionality that is similar (but not recursive) caches the result:

https://github.com/ember-cli/ember-cli/blob/4c60319dfc89234bf95f04a3533ac82bc4246b94/lib/utilities/find-addon-by-name.js#L42

This caching is done as this can be quite costly with many add-ons installed, as this algorithm is recursive it can potentially be even most costly.

To ensure this cache is safe, caching these results per `project` should address any concern.

Reply:
I suspect we shouldn't return all addons, rather simply return true/false if we have duplicates or not (easier to cache, and would be a faster implementation as it wouldn't need to be exhaustive)